shadow-4.0.3/man/mkpasswd.8
.\"$Id: mkpasswd.8 6 2005-03-20 15:34:28Z bubulle $
.\" Copyright 1991, Julianne Frances Haugh
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Julianne F. Haugh nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH MKPASSWD 1
.SH NAME
mkpasswd \- Update passwd and group database files
.SH SYNOPSIS
\fBmkpasswd\fR [\fB-fvgps\fR] \fIfile\fR
.SH DESCRIPTION
.B mkpasswd
reads the file in the format given by the flags and converts it to the
corresponding database file format.
These database files are used to improve access performance on systems
with large numbers of users.
The output files will be named \fIfile\fR.dir and \fIfile\fR.pag.
.PP
The \fB-f\fR option causes \fBmkpasswd\fR to ignore any existing output
files and overwrite them.
Normally \fBmkpasswd\fR complains about existing output files and quits.
.PP
The \fB-v\fR option causes \fBmkpasswd\fR to output information about
each record as it is converted, with a final message at the very end.
.PP
The \fB-g\fR option treats the input file as though it were in
\fI/etc/group\fR file format.
When combined with the \fB-s\fR option, the \fI/etc/gshadow\fR file
format is used instead.
.PP
The \fB-p\fR option treats the input file as though it were in
\fI/etc/passwd\fR file format.
This is the default.
When combined with the \fB-s\fR option, the \fI/etc/shadow\fR file
format is used instead.
.SH CAVEATS
The use of more than one database file is limited to systems which
include the NDBM database library and therefore may not be available
on every system.
.SH NOTE
Since most commands are capable of updating the database files as
changes are made, this command need only be used when re-creating a
deleted or corrupted database file.
.SH FILES
/etc/passwd \- user account information
.br
/etc/shadow \- shadow user information
.br
/etc/group \- group information
.br
/etc/gshadow \- shadow group information
.SH SEE ALSO
.BR group (5),
.BR passwd (5),
.BR shadow (5)
.SH AUTHOR
Julianne Frances Haugh (jockgrrl@ix.netcom.com)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



shadow-4.0.3/src/mkpasswd.c
/*
 * Copyright 1990 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#include "rcsid.h"
RCSID (PKG_VER "$Id: mkpasswd.c 6 2005-03-20 15:34:28Z bubulle $")
#include <sys/stat.h>
#include "prototypes.h"
#include "defines.h"
#include <stdio.h>
#if !defined(NDBM)		/*{ */
int main (int argc, char **argv)
{
	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	fprintf (stderr,
		 _
		 ("%s: no DBM database on system - no action performed\n"),
		 argv[0]);
	return 0;
}

#else				/*} defined(NDBM) { */
#include <fcntl.h>
#include <pwd.h>
#include <ndbm.h>
#include <grp.h>
extern DBM *pw_dbm;
extern DBM *gr_dbm;

#ifdef	SHADOWPWD
extern DBM *sp_dbm;
#endif
#ifdef	SHADOWGRP
extern DBM *sg_dbm;
#endif
char *fgetsx ();

char *Progname;
int vflg = 0;
int fflg = 0;
int gflg = 0;
int sflg = 0;			/* -s flag -- leave in, makes code nicer */
int pflg = 0;

extern struct passwd *sgetpwent ();
extern int pw_dbm_update ();

extern struct group *sgetgrent ();
extern int gr_dbm_update ();

#ifdef	SHADOWPWD
extern struct spwd *sgetspent ();
extern int sp_dbm_update ();
#endif

#ifdef	SHADOWGRP
extern struct sgrp *sgetsgent ();
extern int sg_dbm_update ();
#endif

/* local function prototypes */
static void usage (void);

/*
 * mkpasswd - create DBM files for /etc/passwd-like input file
 *
 * mkpasswd takes an an argument the name of a file in /etc/passwd format
 * and creates a DBM file keyed by user ID and name. The output files have
 * the same name as the input file, with .dir and .pag appended.
 *
 * this command will also create look-aside files for
 * /etc/group, /etc/shadow, and /etc/gshadow.
 */

int main (int argc, char **argv)
{
	extern int optind;
	extern char *optarg;
	FILE *fp;		/* File pointer for input file       */
	char *file;		/* Name of input file                */
	char *dir;		/* Name of .dir file                 */
	char *pag;		/* Name of .pag file                 */
	char *cp;		/* Temporary character pointer       */
	int flag;		/* Flag for command line option      */
	int cnt = 0;		/* Number of entries in database     */
	int longest = 0;	/* Longest entry in database         */
	int len;		/* Length of input line              */
	int errors = 0;		/* Count of errors processing file   */
	char buf[BUFSIZ * 8];	/* Input line from file              */
	struct passwd *passwd = NULL;	/* Pointer to password file entry    */

	struct group *group = NULL;	/* Pointer to group file entry       */

#ifdef	SHADOWPWD
	struct spwd *shadow = NULL;	/* Pointer to shadow passwd entry    */
#endif
#ifdef	SHADOWGRP
	struct sgrp *gshadow = NULL;	/* Pointer to shadow group entry     */
#endif
	DBM *dbm;		/* Pointer to new NDBM files         */
	DBM *dbm_open ();	/* Function to open NDBM files       */

	/*
	 * Figure out what my name is. I will use this later ...
	 */

	Progname = Basename (argv[0]);

	/*
	 * Figure out what the flags might be ...
	 */

	while ((flag = getopt (argc, argv, "fvpgs")) != EOF) {
		switch (flag) {
		case 'v':
			vflg++;
			break;
		case 'f':
			fflg++;
			break;
		case 'g':
			gflg++;
#ifndef	SHADOWGRP
			if (sflg)
				usage ();
#endif
			if (pflg)
				usage ();

			break;
#if defined(SHADOWPWD) || defined(SHADOWGRP)
		case 's':
			sflg++;
#ifndef	SHADOWGRP
			if (gflg)
				usage ();
#endif
			break;
#endif
		case 'p':
			pflg++;
			if (gflg)
				usage ();

			break;
		default:
			usage ();
		}
	}

	/*
	 * Backwards compatibility fix for -p flag ...
	 */

#ifdef SHADOWPWD
	if (!sflg && !gflg)
#else
	if (!gflg)
#endif
		pflg++;

	/*
	 * The last and only remaining argument must be the file name
	 */

	if (argc - 1 != optind)
		usage ();

	file = argv[optind];

	if (!(fp = fopen (file, "r"))) {
		fprintf (stderr, _("%s: cannot open file %s\n"), Progname,
			 file);
		exit (1);
	}

	/*
	 * Make the filenames for the two DBM files.
	 */

	dir = xmalloc (strlen (file) + 5);	/* space for .dir file */
	strcat (strcpy (dir, file), ".dir");

	pag = xmalloc (strlen (file) + 5);	/* space for .pag file */
	strcat (strcpy (pag, file), ".pag");

	/*
	 * Remove existing files if requested.
	 */

	if (fflg) {
		(void) unlink (dir);
		(void) unlink (pag);
	}

	/*
	 * Create the two DBM files - it is an error for these files
	 * to have existed already.
	 */

	if (access (dir, F_OK) == 0) {
		fprintf (stderr, _("%s: cannot overwrite file %s\n"),
			 Progname, dir);
		exit (1);
	}
	if (access (pag, F_OK) == 0) {
		fprintf (stderr, _("%s: cannot overwrite file %s\n"),
			 Progname, pag);
		exit (1);
	}

	if (sflg)
		umask (077);
	else
		umask (022);

	/*
	 * Now the DBM database gets initialized
	 */

	if (!(dbm = dbm_open (file, O_RDWR | O_CREAT, 0644))) {
		fprintf (stderr, _("%s: cannot open DBM files for %s\n"),
			 Progname, file);
		exit (1);
	}
	if (gflg) {
#ifdef	SHADOWGRP
		if (sflg)
			sg_dbm = dbm;
		else
#endif
			gr_dbm = dbm;
	} else {
#ifdef	SHADOWPWD
		if (sflg)
			sp_dbm = dbm;
		else
#endif
			pw_dbm = dbm;
	}

	/*
	 * Read every line in the password file and convert it into a data
	 * structure to be put in the DBM database files.
	 */

	while (fgetsx (buf, BUFSIZ, fp) != NULL) {

		/*
		 * Get the next line and strip off the trailing newline
		 * character.
		 */

		buf[sizeof buf - 1] = '\0';
		if (!(cp = strchr (buf, '\n'))) {
			fprintf (stderr,
				 _("%s: the beginning with " %
				   .16 s..." is too long\n"), Progname,
				 buf);
			exit (1);
		}
		*cp = '\0';
		len = strlen (buf);

#ifdef	USE_NIS
		/*
		 * Parse the password file line into a (struct passwd). 
		 * Erroneous lines cause error messages, but that's all. YP
		 * lines are ignored completely.
		 */

		if (buf[0] == '-' || buf[0] == '+')
			continue;
#endif
		if (!(((!sflg && pflg) && (passwd = sgetpwent (buf)))
#ifdef	SHADOWPWD
		      || ((sflg && pflg) && (shadow = sgetspent (buf)))
#endif
		      || ((!sflg && gflg) && (group = sgetgrent (buf)))
#ifdef	SHADOWGRP
		      || ((sflg && gflg) && (gshadow = sgetsgent (buf)))
#endif
		    )) {
			fprintf (stderr,
				 _("%s: error parsing line \"%s\"\n"),
				 Progname, buf);
			errors++;
			continue;
		}
		if (vflg) {
			if (!sflg && pflg)
				printf (_
					("adding record for name " %
					 s "\n"), passwd->pw_name);
#ifdef	SHADOWPWD
			if (sflg && pflg)
				printf (_
					("adding record for name " %
					 s "\n"), shadow->sp_namp);
#endif
			if (!sflg && gflg)
				printf (_
					("adding record for name " %
					 s "\n"), group->gr_name);
#ifdef	SHADOWGRP
			if (sflg && gflg)
				printf (_
					("adding record for name " %
					 s "\n"), gshadow->sg_name);
#endif
		}
		if (!sflg && pflg && !pw_dbm_update (passwd))
			fprintf (stderr,
				 _("%s: error adding record for " %
				   s "\n"), Progname, passwd->pw_name);

#ifdef	SHADOWPWD
		if (sflg && pflg && !sp_dbm_update (shadow))
			fprintf (stderr,
				 _("%s: error adding record for " %
				   s "\n"), Progname, shadow->sp_namp);
#endif
		if (!sflg && gflg && !gr_dbm_update (group))
			fprintf (stderr,
				 _("%s: error adding record for " %
				   s "\n"), Progname, group->gr_name);
#ifdef	SHADOWGRP
		if (sflg && gflg && !sg_dbm_update (gshadow))
			fprintf (stderr,
				 _("%s: error adding record for " %
				   s "\n"), Progname, gshadow->sg_name);
#endif				/* SHADOWGRP */

		/*
		 * Update the longest record and record count
		 */

		if (len > longest)
			longest = len;
		cnt++;
	}

	/*
	 * Tell the user how things went ...
	 */

	if (vflg)
		printf (_("added %d entries, longest was %d\n"), cnt,
			longest);

	exit (errors);
 /*NOTREACHED*/}

/*
 * usage - print error message and exit
 */

static void usage (void)
{
#ifdef	SHADOWPWD
#ifdef	SHADOWGRP
	fprintf (stderr, _("Usage: %s [-vf] [-p|g|sp|sg] file\n"),
		 Progname);
#else				/* !SHADOWGRP */
	fprintf (stderr, _("Usage: %s [-vf] [-p|g|sp] file\n"), Progname);
#endif				/* SHADOWGRP */
#else				/* !SHADOWPWD */
	fprintf (stderr, _("Usage: %s [-vf] [-p|g] file\n"), Progname);
#endif				/* SHADOWPWD */

	exit (1);
 /*NOTREACHED*/}
#endif				/*} defined(NDBM) */
